### PR TITLE
Fix k8s pod.execute randomly stuck indefinitely by logs consumption (#23497)

### PR DESCRIPTION
**Pull request by schattian**
related: #23618 
fixes leaking loop in https://github.com/apache/airflow/commit/67474cccb612fad39a2afbeee16014909447d242. That also involved changing `get_event_loop` to be `new_event_loop` so it is closable (otherwise, parallel usage of the function could try to reuse a closed loop). 

Finally, it fixes previously broken (modified) test. Notice that one call to container_is_running was leaked to one test that only made 3 itself.

Unfortunately, cant run all ci tests setup in my machine (oom)

### CODE REVIEW

**Comment by schattian** [comment]
Ok, i think i figured out the problem. `shutdown_default_executor` is not available until py3.9.
Therefore i need to create an executor instead of rely in the default one (notice `loop.close` does not wait until thread of executor finishes)

**Comment by potiuk** [comment]
Yeah. Using Asyncio is kinda tricky for standalone operator. It sounds a bit tricky with this approach. Seems that we are trying to start asyncio.loop to run asyncio operations and then we kill the thread after closing the operator. This sounds rather complex as we do not really achieve the "async" promises.

Just a thought - maybe we should somehew plug it in the "defferable" operators and triggerer ? I am not sure if that would be efficient for getting the logs using triggerer and triggers ? Or maybe we should abandon the idea of async operations here as they do not really give us anything and only cause troublesome testing and difficult debugging in the future?

@andrewgodwin  - maybe you can share your thoughts on that ?

**Comment by schattian** [comment]
@potiuk I had one thing in mind to try to reduce the complexity of this, releasing the connection (response.release_conn). Based on https://github.dev/urllib3/urllib3/blob/37a67739ba9a7835886555744e9dabe3009e538a/src/urllib3/response.py#L708-L712 seems could work. 

Much harder to test though as rn is just a simple generator.

**Comment by schattian** [comment]
@potiuk i replaced the asyncio-based approach with an executor to use multiprocessing.Process instead.
I added one more test case also for the case api hangs + no logs.
Imho this is the best to do to not overcomplicate this, wdyt?

File: kubernetes_tests/test_kubernetes_pod_operator.py [hunk 1]
```diff
@@ -376,7 +377,9 @@ def test_port(self):
         assert self.expected_pod == actual_pod
 
     def test_volume_mount(self):
-        with mock.patch.object(PodManager, 'log') as mock_logger:
+        with mock.patch.object(
```
Comment (schattian) [inline kubernetes_tests/test_kubernetes_pod_operator.py:434]: notice a non-thread-safe counter (default) makes this test flaky or impossible to do as the call of log.info happens in a process.

**Comment by schattian** [comment]
@potiuk could you please take a look at this one? would be nice to have, personally I experience this bug regularly (once a week) on gke.

**Review by potiuk — APPROVED** [review]
LGTM. But I would love @dstandish or  @jedcunningham or @ephraimbuddy to take a look too. Maybe there is a simpler way to approach it. Sounds like this should be handled by k8s library actually.

**Comment by potiuk** [comment]
(And Sorry for long delays in review @schattian - we had Airflow Summit last week which kept many of the reviewers busy).

**Comment by schattian** [comment]
> (And Sorry for long delays in review @schattian - we had Airflow Summit last week which kept many of the reviewers busy).

didnt know that. no worries! and thanks for the rev

**Comment by potiuk** [comment]
@dstandish @ashb @ephraimbuddy @jedcunningham - I think we need some of your's expert knowledge here

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py [hunk 2]
```diff
@@ -256,7 +293,7 @@ def consume_logs(*, since_time: Optional[DateTime] = None, follow: bool = True)
                 time.sleep(1)
 
     def await_container_completion(self, pod: V1Pod, container_name: str) -> None:
-        while not self.container_is_running(pod=pod, container_name=container_name):
+        while self.container_is_running(pod=pod, container_name=container_name):
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/utils/pod_manager.py]: this looks like it is unrelated to your main change.  could you please raise a separatae PR for this change specifically?

Comment (dstandish) [reply airflow/providers/cncf/kubernetes/utils/pod_manager.py]: (it's less complicated, more obvious fix than the other part  so we can merge it faster and focus more cleanly on the other bit in this PR)

**Comment by dstandish** [comment]
@schattian any reason you are using multiprocessing instead of a thread?

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py [hunk 3]
```diff
@@ -193,6 +195,40 @@ def follow_container_logs(self, pod: V1Pod, container_name: str) -> PodLoggingSt
...
+            self.log.info(message)
+        return timestamp
+
+    def consume_container_logs_stream(
+        self, pod: V1Pod, container_name: str, stream: Iterable[bytes]
+    ) -> Optional[DateTime]:
+        def log_iterable_and_set_value(timestamp):
+            dt = self.log_iterable(stream)
+            if dt is not None:
+                timestamp.value = dt.timestamp()  # type: ignore[attr-defined]
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/utils/pod_manager.py:211]: shouldn't this be `timestamp.value = dt`? ultimately this gets returned through `consume_logs` which returns pendulum.DateTime

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py (hunk 2 above)
Comment (dstandish) [reply airflow/providers/cncf/kubernetes/utils/pod_manager.py]: ah, done here: https://github.com/apache/airflow/pull/23883/files now just need to rebase

Comment (schattian) [reply airflow/providers/cncf/kubernetes/utils/pod_manager.py]: nice - yep, i though on doing so in a diff pr but im using that func.
Will rebase

**Comment by schattian** [comment]
> @schattian any reason you are using multiprocessing instead of a thread?
Yep. The thread could hang indefinitely, as the kube logs api stream could do, so I need to terminate the process.

However, i am not very knowledgeable on python's api for this kind of thing, so open to suggestions.

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py (hunk 3 above)
Comment (schattian) [reply airflow/providers/cncf/kubernetes/utils/pod_manager.py:211]: i cannot do it with dt as i am using a rawvalue('f') -float-. Therefore, i need to use the unix ts instead to pass it.

**Comment by potiuk** [comment]
> > @schattian any reason you are using multiprocessing instead of a thread?
> 
> @dstandish Yep. The thread could hang indefinitely, as the kube logs api stream could do, so I need to terminate the process.
> 
> However, i am not very knowledgeable on python's api for this kind of thing, so open to suggestions.

OK. I will wait for merging before releasing the new wave of providers then - I think it might need some more deliberation (We will include it in the next wave then).

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py (hunk 2 above)
Comment (potiuk) [reply airflow/providers/cncf/kubernetes/utils/pod_manager.py]: It looks green. @dstandish - do you approve :)?

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py (hunk 3 above)
Comment (dstandish) [reply airflow/providers/cncf/kubernetes/utils/pod_manager.py:211]: ok but when this value is returned by consume_logs, you must ensure that it is converted back to DateTime right? perhaps you have done that somewhere?

Comment (dstandish) [reply airflow/providers/cncf/kubernetes/utils/pod_manager.py:211]: nevermind i see it now

**Comment by dstandish** [comment]
also, i'd make "private" the two methods you've added

**Comment by dstandish** [comment]
separate question for you @schattian. suppose the stream hangs, and the pod is still running.  if we terminate the process and restart logs consumption, will it get logs from the new location?  or will it still look at the original location and therefore any new logging is unreachable and it will immediately hang?

**Comment by schattian** [comment]
>one alternative i'll just mention is we could handle this by doing our own "following" instead of relying on the stream. i.e. we do our own loop and consume the latest logs, then sleep a second and do the same.

you mean calling X amount of times the kubectl logs api? I think that's a worse than a follow.. 
Mostly because you'll have some lines printed twice (and could be expensive for some dags)

>one is I think you need to kill the logging process when on_kill is called.

Sorry, i dont think i understand what's the point here

>the other is, if you get an error in await_container_completion, your subprocess won't be killed. so, you probably should modify so that if you get an unexpected error after process has started, that the process will be killed.

Hmm but in that case, aren't we going to propagate a SIGTERM? (ie killing the subprocess implicitly)?
I was just relying on that, but if that's not how it works / could change, i will modify that.

>also, i'd make "private" the two methods you've added
👍🏻 

>separate question for you @schattian. suppose the stream hangs, and the pod is still running. if we terminate the process and restart logs consumption, will it get logs from the new location? or will it still look at the original location and therefore any new logging is unreachable and it will immediately hang?

That is correct. A call to kubectl logs api after the stream hanged will return the newest logs.
But, there are a few problems with that:

1. The pod could be deleted, and in that case you'll receive no logs.
2. We determine that the stream hangs as "if the container is not running and stream consumption didnt finish", as there is not a better way to do it (please let me know if you have one). In that case

I have mentioned that in the first pr https://github.com/apache/airflow/pull/23618#discussion_r869096030
I think it adds some complexity that's not worth as this bugs doesn't occur so often

**Comment by dstandish** [comment]
>> one is I think you need to kill the logging process when on_kill is called.

>Sorry, i dont think i understand what's the point here

in operators, if we kill the task (e.g. from UI) shouldn't we also kill the logging subprocess?  if we don't do that, will it continue running e.g. on celery worker?

**Comment by schattian** [comment]
> in operators, if we kill the task (e.g. from UI) shouldn't we also kill the logging subprocess? if we don't do that, will it continue running e.g. on celery worker?
Ah, okay, yeah that's what i was trying to answer with:

>Hmm but in that case, aren't we going to propagate a SIGTERM? (ie killing the subprocess implicitly)?
>I was just relying on that, but if that's not how it works / could change, i will modify that.

wdyt?

**Comment by schattian** [comment]
@dstandish i've forgot, but what happens in such case (and same for whenever the pod is deleted for some other external issue that could cause, e.g: pod not found) is that the kubectl stream api returns immediately, so this is not an issue (see eg cncf kube pod operator).

I've just made the functions private

**Comment by schattian** [comment]
hm, are calls to method wrapped by a private function mocked?
I dont see any other way the tests can timeout with the new changes. They are not flaky, only fail consistently after the addition

**Comment by potiuk** [comment]
@schattian ALWAYS rebase your changes when you attempt to see if the tests are passing. This way you avpod wasting your and committers time on trying to solve problems that are Likely already solved. 

You are 102 commits behind. Things ARE not going to work.

image

**Comment by potiuk** [comment]
> @potiuk uhm i dont see how it helps. the ci code was the same as before except the new additions, and tests fail for those.
> 
> I've updated it btw, lets see

I think this is pretty obvious - if there was an issue with some tests at the moment you branched off no matter how hard you try it will stay there if you don't rebase - if it was fixed in one of the 102 commits you will get it fixed.
Why do you think it would work otherwise?

**Comment by potiuk** [comment]
Now - it seem  you do have a real issue - lilely introduced by your changes:

You need to (as usual) start breeze in your version (build image) -  reproduce it by running tests in the same environment as in CI and fix.

**Comment by potiuk** [comment]
> I think this is pretty obvious - if there was an issue with some tests at the moment you branched off no matter how hard you try it will stay there if you don't rebase - if it was fixed in one of the 102 commits you will get it fixed.

BTW. It's not guaranteed to work. It just removes one of the potential failure reasons out of the possible issues.

**Comment by schattian** [comment]
yeah thats what i was thinking, added by making the functions private.
What i think happens is that the methods that are looking to be mocked are inside a private function.
I dont see how to change that behaviour other than making the functions non-private again.

**Comment by dstandish** [comment]
> yeah thats what i was thinking, added by making the functions private.
What i think happens is that the methods that are looking to be mocked are inside a private function.
I dont see how to change that behaviour other than making the functions non-private again

wellp sorry yeah if they can't be private they can't be private.  i just want to minimize surface area of promises we are making.   when we have "public" methods we can't change them without deprecation.

re 

> @dstandish i've forgot, but what happens in such case (and same for whenever the pod is deleted for some other external issue that could cause, for example, pod not found) is that the kubectl stream api returns immediately, so this is not an issue (see eg cncf kube pod operator on_kill def).

can you clarify...  when _what_ happens?

**Comment by schattian** [comment]
> can you clarify...  when _what_ happens?

The function consuming the logs (the process) will return immediately so it'll terminate.

That's because the stream api will return.

**Comment by dstandish** [comment]
> > can you clarify...  when _what_ happens?
> 
> The function consuming the logs (the process) will return immediately so it'll terminate.
> 
> That's because the stream api will return.

Oh i think i understand what you're saying.  You are saying that, since on_kill terminates the pod, the log stream will close and the process will die of its own accord, so there's no need to manage the termination of that process directly.

**Pull request closed without merging**